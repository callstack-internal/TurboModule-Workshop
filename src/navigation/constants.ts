export const Screens = {
  Main: {
    title: 'Catálogo de Perros',
    name: 'Main',
  },
  HowItWorks: {
    title: 'Cómo funciona',
    name: 'HowItWorks',
  },
  Home: {
    title: 'Home',
    name: 'Home',
  },
  DogDetails: {
    title: 'Detalles',
    name: 'DogDetails',
  },
  Adoptar: {
    title: 'Adoptar',
    name: 'Adoptar',
  },
  MisPerros: {
    title: 'Mis Perros',
    name: 'MisPerros',
  },
} as const;
