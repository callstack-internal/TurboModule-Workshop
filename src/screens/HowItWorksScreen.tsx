import React from 'react';
import {
  View,
  Text,
  StyleSheet,
  ScrollView,
  SafeAreaView,
  StatusBar,
  TouchableOpacity,
} from 'react-native';
import { useNavigation } from '@react-navigation/native';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons';

function HowItWorksScreen() {
  const navigation = useNavigation();

  return (
    <SafeAreaView style={styles.container}>
      <StatusBar barStyle="dark-content" backgroundColor="#f8f9fa" />

      <View style={styles.header}>
        <View style={styles.headerRow}>
          <TouchableOpacity style={styles.menuButton} onPress={() => navigation.goBack()}>
            <Icon name="arrow-left" size={24} color="#212529" />
          </TouchableOpacity>
          <Text style={styles.title}>¿Cómo funciona la adopción?</Text>
        </View>
      </View>

      <ScrollView contentContainerStyle={styles.scrollContent}>
        <View style={styles.section}>
          <Text style={styles.sectionTitle}>1. Explora nuestro catálogo</Text>
          <Text style={styles.sectionText}>
            En la pestaña "Adoptar" encontrarás todos los perros disponibles para adopción. Puedes
            buscar por nombre y ordenarlos por cercanía a tu ubicación.
          </Text>
        </View>

        <View style={styles.section}>
          <Text style={styles.sectionTitle}>2. Conoce a los perros</Text>
          <Text style={styles.sectionText}>
            Toca en cualquier perro para ver más detalles sobre él. Podrás ver su edad, raza,
            ubicación y si está disponible para adopción.
          </Text>
        </View>

        <View style={styles.section}>
          <Text style={styles.sectionTitle}>3. Adopta un perro</Text>
          <Text style={styles.sectionText}>
            Si un perro está disponible para adopción, verás un botón "Adoptar" en su página de
            detalles. Al presionarlo, confirmarás tu intención de adoptar.
          </Text>
        </View>

        <View style={styles.section}>
          <Text style={styles.sectionTitle}>4. Gestiona tus adopciones</Text>
          <Text style={styles.sectionText}>
            En la pestaña "Mis Perros" podrás ver todos los perros que has adoptado. Esta lista se
            guarda automáticamente, así que no perderás tus adopciones aunque cierres la aplicación.
          </Text>
        </View>

        <View style={styles.section}>
          <Text style={styles.sectionTitle}>Perros no disponibles</Text>
          <Text style={styles.sectionText}>
            Algunos perros no están disponibles para adopción inmediata. Esto puede deberse a que
            están enfermos o son muy jóvenes. En estos casos, verás la fecha a partir de la cual
            estarán disponibles.
          </Text>
        </View>
      </ScrollView>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f8f9fa',
  },
  scrollContent: {
    paddingBottom: 24,
  },
  header: {
    padding: 16,
    backgroundColor: '#ffffff',
    borderBottomWidth: 1,
    borderBottomColor: '#e9ecef',
    elevation: 2,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.1,
    shadowRadius: 3,
  },
  headerRow: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  menuButton: {
    marginRight: 16,
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
    color: '#212529',
    flex: 1,
  },
  section: {
    backgroundColor: '#ffffff',
    marginHorizontal: 16,
    marginTop: 16,
    padding: 16,
    borderRadius: 8,
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 0.05,
    shadowRadius: 2,
    elevation: 1,
  },
  sectionTitle: {
    fontSize: 18,
    fontWeight: 'bold',
    color: '#212529',
    marginBottom: 8,
  },
  sectionText: {
    fontSize: 16,
    color: '#495057',
    lineHeight: 24,
  },
});

export default HowItWorksScreen;
