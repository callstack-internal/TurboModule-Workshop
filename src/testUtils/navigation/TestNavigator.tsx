import React, { useEffect } from 'react';

import {
  createNavigatorFactory,
  useNavigationBuilder,
  StackRouter,
  TabRouter,
  DrawerRouter,
  DefaultRouterOptions,
  StackRouterOptions,
  TabRouterOptions,
  DrawerRouterOptions,
  NavigationHelpers,
  ParamListBase,
  RouterFactory,
  Router,
  NavigationState,
  NavigationAction,
  NavigationContainer,
} from '@react-navigation/native';
import type {
  StackNavigationState,
  TabNavigationState,
  DrawerNavigationState,
} from '@react-navigation/native';
import type {
  StackActionHelpers,
  TabActionHelpers,
  DrawerActionHelpers,
} from '@react-navigation/native';
import { Text } from 'react-native';

type TestScreenOptions = Record<string, unknown>;

type NavigatorType = 'stack' | 'tabs' | 'drawer';

interface TestNavigatorImplProps extends DefaultRouterOptions {
  type: NavigatorType;
  screenOptions?: TestScreenOptions;
  initialRouteName?: string;
  children: React.ReactNode;
}

type AnyNavigationState =
  | StackNavigationState<ParamListBase>
  | TabNavigationState<ParamListBase>
  | DrawerNavigationState<ParamListBase>;

type AnyRouterOptions = StackRouterOptions | TabRouterOptions | DrawerRouterOptions;

type AnyActionHelpers =
  | StackActionHelpers<ParamListBase>
  | TabActionHelpers<ParamListBase>
  | DrawerActionHelpers<ParamListBase>;

type ExtractedAction = Exclude<
  Parameters<Router<NavigationState, NavigationAction>['getStateForAction']>[0],
  undefined
>;

/**
 * Custom navigator component used for testing React Navigation screens.
 * Supports `stack`, `tabs`, and `drawer` types.
 * Renders all screens at once and injects a `focused` prop based on navigation state.
 *
 * Useful in tests to simulate navigation behavior with a lightweight navigator.
 */
const TestNavigatorImpl = ({ type, ...rest }: TestNavigatorImplProps): React.ReactElement => {
  const router = (
    type === 'drawer' ? DrawerRouter : type === 'stack' ? StackRouter : TabRouter
  ) as RouterFactory<AnyNavigationState, ExtractedAction, AnyRouterOptions>;

  const {
    state,
    descriptors,
    NavigationContent: Content,
  } = useNavigationBuilder<
    AnyNavigationState,
    AnyRouterOptions,
    AnyActionHelpers,
    TestScreenOptions,
    NavigationHelpers<ParamListBase>
  >(router, rest);

  return <Content>{state.routes.map((route) => descriptors[route.key]?.render())}</Content>;
};

/**
 * Creates a navigator factory for the TestNavigator.
 * This wraps the custom `TestNavigator` using `createNavigatorFactory`,
 * enabling use of `Navigator` and `Screen` components just like built-in navigators.
 *
 * Use this when integrating `TestNavigator` into `NavigationContainer`.
 */
const createTestNavigatorFactory = createNavigatorFactory(TestNavigatorImpl);

type LayoutProps = {
  name: string;
  focused: boolean;
  navigation: NavigationHelpers<ParamListBase>;
  children: React.ReactNode;
};

const CustomLayout = ({ name, focused, navigation, children }: LayoutProps): React.ReactElement => {
  useEffect(() => {
    const state = navigation.getState();
    const currentRoute = state.routes[state.index];

    if (focused && currentRoute?.name !== name) {
      navigation.navigate(name);
    } else if (!focused && currentRoute?.name !== 'Dummy') {
      navigation.navigate('Dummy');
    }
  }, [focused, name, navigation]);

  return <>{children}</>;
};

const TestLayout = ({ name, focused, ...props }: LayoutProps): React.ReactElement => {
  return <CustomLayout {...props} name={name} focused={focused} />;
};

export const TestNavigator = ({
  children,
  type = 'stack',
  name = 'TestScreen',
  focused = true,
  screenParams,
  dummyName = 'Dummy',
}: {
  children: React.ReactNode;
  type?: NavigatorType;
  name?: string;
  focused?: boolean;
  screenParams?: object;
  dummyName?: string;
}) => {
  const Test = createTestNavigatorFactory();

  return (
    <NavigationContainer>
      <Test.Navigator
        key={type}
        type={type}
        initialRouteName={focused ? name : dummyName}
        layout={
          /* eslint-disable-next-line react/no-unstable-nested-components */
          (props: Omit<LayoutProps, 'name' | 'focused'>) => (
            <TestLayout {...props} name={name} focused={focused} />
          )
        }
      >
        <Test.Screen name={name} getId={() => name} initialParams={screenParams}>
          {() => children}
        </Test.Screen>
        <Test.Screen name={dummyName} getId={() => dummyName}>
          {() => <Text testID="dummy-screen">Dummy</Text>}
        </Test.Screen>
      </Test.Navigator>
    </NavigationContainer>
  );
};
