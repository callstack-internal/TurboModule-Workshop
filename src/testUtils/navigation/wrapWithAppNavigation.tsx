import React from 'react';

import {
  InitialState,
  NavigationContainer,
  NavigationContainerRef,
} from '@react-navigation/native';
import DrawerNavigator from '../../navigation/navigators/DrawerNavigator.tsx';

type WrapWithAppNavigationProps = {
  initialState?: InitialState;
  navigationRef?: React.RefObject<NavigationContainerRef<any>>;
};

/**
 * Wraps the full app navigator (`DrawerNavigator`) in a `NavigationContainer` for testing.
 *
 * This is the preferred utility when writing tests for **real screen components or components that navigate** that are part of
 * the actual app navigation tree. It gives you a full, realistic navigation environment without mocks,
 * including nested navigators, screen options, transitions, etc.
 *
 * Unlike `TestNavigator`, this uses your actual app's navigator configuration, making it ideal for testing:
 * - Navigation behavior (e.g., `navigate`, `goBack`, deep linking)
 * - Screen lifecycle hooks (e.g., `useFocusEffect`, `useNavigation`)
 * - Integration-level rendering of screen flows
 *
 * ⚠️ Note: This should be used when testing **screen components**.
 * If you're only testing a non-screen component (e.g. custom hook, modal, header)
 * that requires navigation context, consider using `TestNavigator` or `createTestNavigator` instead for a minimal setup.
 *
 */
export const wrapWithAppNavigation = ({
  initialState,
  navigationRef,
}: WrapWithAppNavigationProps) => {
  return (
    <NavigationContainer initialState={initialState} ref={navigationRef}>
      <DrawerNavigator />
    </NavigationContainer>
  );
};
