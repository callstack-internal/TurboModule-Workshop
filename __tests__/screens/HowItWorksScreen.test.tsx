import { render, fireEvent, waitFor, act } from '@testing-library/react-native';
import { Screens } from '../../src/navigation/constants.ts';
import { wrapWithAppNavigation } from '../../src/testUtils/navigation/wrapWithAppNavigation.tsx';
import { TestNavigator } from '../../src/testUtils/navigation/TestNavigator.tsx';
import HowItWorksScreen from '../../src/screens/HowItWorksScreen';

const initialState = {
  routes: [
    {
      name: Screens.Main.name,
      state: {
        index: 1,
        routes: [
          {
            name: Screens.Home.name,
          },
          {
            name: Screens.HowItWorks.name,
          },
        ],
      },
    },
  ],
};

const component = () => render(wrapWithAppNavigation({ initialState }));

describe('HowItWorksScreen', () => {
  test('renders correctly with all sections', async () => {
    const { getByText, getByTestId } = component();

    await waitFor(() => {
      expect(getByTestId('back-button')).toBeTruthy();

      // Check if the title is displayed
      expect(getByText('¿Cómo funciona la adopción?')).toBeTruthy();

      // Check if all section titles are displayed
      expect(getByText('1. Explora nuestro catálogo')).toBeTruthy();
      expect(getByText('2. Conoce a los perros')).toBeTruthy();
      expect(getByText('3. Adopta un perro')).toBeTruthy();
      expect(getByText('4. Gestiona tus adopciones')).toBeTruthy();
      expect(getByText('Perros no disponibles')).toBeTruthy();

      // Check if some section text is displayed
      expect(
        getByText(/En la pestaña "Adoptar" encontrarás todos los perros disponibles/),
      ).toBeTruthy();
      expect(getByText(/Toca en cualquier perro para ver más detalles/)).toBeTruthy();
      expect(getByText(/Si un perro está disponible para adopción/)).toBeTruthy();
      expect(getByText(/En la pestaña "Mis Perros" podrás ver todos los perros/)).toBeTruthy();
      expect(getByText(/Algunos perros no están disponibles para adopción inmediata/)).toBeTruthy();
    });
  });

  test('navigates back when back button is pressed', async () => {
    const { getByTestId, getByText } = component();

    await waitFor(() => {
      expect(getByTestId('back-button')).toBeTruthy();
    });

    act(() => fireEvent.press(getByTestId('back-button')));

    await waitFor(() => {
      expect(getByText('Obtener Ubicación')).toBeTruthy();
    });
  });

  test('matches snapshot', async () => {
    const { toJSON } = render(
      <TestNavigator>
        <HowItWorksScreen />
      </TestNavigator>,
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
