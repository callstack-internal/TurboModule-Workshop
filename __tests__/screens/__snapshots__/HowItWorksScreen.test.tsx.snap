// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HowItWorksScreen matches snapshot 1`] = `
<RCTSafeAreaView
  style={
    {
      "backgroundColor": "#f8f9fa",
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#ffffff",
        "borderBottomColor": "#e9ecef",
        "borderBottomWidth": 1,
        "elevation": 2,
        "padding": 16,
        "shadowColor": "#000",
        "shadowOffset": {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 3,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "marginRight": 16,
            "opacity": 1,
          }
        }
        testID="back-button"
      >
        <Icon
          color="#212529"
          name="arrow-left"
          size={24}
        />
      </View>
      <Text
        style={
          {
            "color": "#212529",
            "flex": 1,
            "fontSize": 24,
            "fontWeight": "bold",
          }
        }
      >
        ¿Cómo funciona la adopción?
      </Text>
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      {
        "paddingBottom": 24,
      }
    }
  >
    <View>
      <View
        style={
          {
            "backgroundColor": "#ffffff",
            "borderRadius": 8,
            "elevation": 1,
            "marginHorizontal": 16,
            "marginTop": 16,
            "padding": 16,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.05,
            "shadowRadius": 2,
          }
        }
      >
        <Text
          style={
            {
              "color": "#212529",
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 8,
            }
          }
        >
          1. Explora nuestro catálogo
        </Text>
        <Text
          style={
            {
              "color": "#495057",
              "fontSize": 16,
              "lineHeight": 24,
            }
          }
        >
          En la pestaña "Adoptar" encontrarás todos los perros disponibles para adopción. Puedes buscar por nombre y ordenarlos por cercanía a tu ubicación.
        </Text>
      </View>
      <View
        style={
          {
            "backgroundColor": "#ffffff",
            "borderRadius": 8,
            "elevation": 1,
            "marginHorizontal": 16,
            "marginTop": 16,
            "padding": 16,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.05,
            "shadowRadius": 2,
          }
        }
      >
        <Text
          style={
            {
              "color": "#212529",
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 8,
            }
          }
        >
          2. Conoce a los perros
        </Text>
        <Text
          style={
            {
              "color": "#495057",
              "fontSize": 16,
              "lineHeight": 24,
            }
          }
        >
          Toca en cualquier perro para ver más detalles sobre él. Podrás ver su edad, raza, ubicación y si está disponible para adopción.
        </Text>
      </View>
      <View
        style={
          {
            "backgroundColor": "#ffffff",
            "borderRadius": 8,
            "elevation": 1,
            "marginHorizontal": 16,
            "marginTop": 16,
            "padding": 16,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.05,
            "shadowRadius": 2,
          }
        }
      >
        <Text
          style={
            {
              "color": "#212529",
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 8,
            }
          }
        >
          3. Adopta un perro
        </Text>
        <Text
          style={
            {
              "color": "#495057",
              "fontSize": 16,
              "lineHeight": 24,
            }
          }
        >
          Si un perro está disponible para adopción, verás un botón "Adoptar" en su página de detalles. Al presionarlo, confirmarás tu intención de adoptar.
        </Text>
      </View>
      <View
        style={
          {
            "backgroundColor": "#ffffff",
            "borderRadius": 8,
            "elevation": 1,
            "marginHorizontal": 16,
            "marginTop": 16,
            "padding": 16,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.05,
            "shadowRadius": 2,
          }
        }
      >
        <Text
          style={
            {
              "color": "#212529",
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 8,
            }
          }
        >
          4. Gestiona tus adopciones
        </Text>
        <Text
          style={
            {
              "color": "#495057",
              "fontSize": 16,
              "lineHeight": 24,
            }
          }
        >
          En la pestaña "Mis Perros" podrás ver todos los perros que has adoptado. Esta lista se guarda automáticamente, así que no perderás tus adopciones aunque cierres la aplicación.
        </Text>
      </View>
      <View
        style={
          {
            "backgroundColor": "#ffffff",
            "borderRadius": 8,
            "elevation": 1,
            "marginHorizontal": 16,
            "marginTop": 16,
            "padding": 16,
            "shadowColor": "#000",
            "shadowOffset": {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.05,
            "shadowRadius": 2,
          }
        }
      >
        <Text
          style={
            {
              "color": "#212529",
              "fontSize": 18,
              "fontWeight": "bold",
              "marginBottom": 8,
            }
          }
        >
          Perros no disponibles
        </Text>
        <Text
          style={
            {
              "color": "#495057",
              "fontSize": 16,
              "lineHeight": 24,
            }
          }
        >
          Algunos perros no están disponibles para adopción inmediata. Esto puede deberse a que están enfermos o son muy jóvenes. En estos casos, verás la fecha a partir de la cual estarán disponibles.
        </Text>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
